README.md

# uni-app-cli-preset

[更新记录](./CHANGELOG.md)

**开箱即用！！基于 ```uni-app cli``` 模式集成最佳实践脚手架**

- 集成多平台快速开发的 UI 框架，[uView UI](https://www.uviewui.com) 2.0 版

  `uView UI` 是全面兼容 nvue 的 uni-app 生态框架，全面的组件和便捷的工具会让您信手拈来，如鱼得水

- 集成 [ColorUI](https://github.com/weilanwl/ColorUI)

  `ColorUI` 是一个 Css 类的 UI 组件库！不是一个 Js 框架。相比于同类小程序组件库，ColorUI 更注重于视觉交互！
  
  - 源代码：[GitHub](https://github.com/weilanwl/ColorUI)

  - [ColorUI 使用文档]('http://docs.xzeu.com/#/info/%E5%BF%AB%E9%80%9F%E5%BC%80%E5%A7%8B/%E5%BF%AB%E9%80%9F%E5%B8%83%E7%BD%B2')

  - [ColorUI群资源](https://www.yuque.com/colorui)

  **注：本脚手架集成的 `color-ui` 并不是原作者的源。**
  
    > 为了适应项目的开发需求，基于 `color-ui` 进行二次开发的 sass 版本 `color-ui-sass`
      其功能上保持不变。

- 集成 [luch-request](https://github.com/lei-mu/luch-request)
  
  `luch-request` 是一个基于 Promise 开发的 uni-app 跨平台、项目级别的请求库，它有更小的体积，易用的api，方便简单的自定义能力。

  - 源代码：[GitHub](https://github.com/lei-mu/luch-request)

  - [官网&使用文档]('https://www.quanzhan.co/luch-request/')

- 集成当前主流的三种预编译器：`Less` `SCSS` `Stylus`
  
  CSS 预处理器是用一种专门的编程语言，进行 Web 页面样式设计，然后再编译成正常的 CSS 文件，以供项目使用。让你的 CSS 更加简洁、适应性更强、可读性更佳，更易于代码的维护等诸多好处。

- 集成 `Pug` Template

  pug 是一种前端模板引擎，原名 jade 可用来生成 HTML，它的写法类似于 CSS，语法简单易理解，同时极大的简约了我们的代码。
  
  **WHY PUG?**
  > As web designers or developers, we likely all have to write our fair share of HTML. And while this is not the most difficult task, it can often feel a little boring or repetitive. This is where the Pug HTML preprocessor comes in.
  
  [中文文档](https://pugjs.org/zh-cn/api/getting-started.html)

- 集成 z-paging

  一个 uni-app 分页组件。

  全平台兼容，支持自定义下拉刷新、上拉加载更多，支持虚拟列表，支持自动管理空数据图、点击返回顶部，支持聊天分页、本地分页，支持展示最后更新时间，支持国际化等等。

  - 源代码：[GitHub](https://github.com/SmileZXLee/uni-z-paging)
  - [官网&使用文档](https://z-paging.zxlee.cn/start/intro.html)  

- 集成项目规范（eslint + prettier）支持 `standard` 和 `eslint:recommended`,默认配置 `standard`

## 上手指南

### 安装依赖

```bash
yarn
```

### 运行项目

```bash
# 默认为H5环境 更多环境命令请参考 package.json 中的 scripts字段
yarn serve 
```

### 打包发行

```bash
yarn build # 默认为H5环境 更多环境命令请参考 package.json 中的 scripts字段
```

## 微信小程序

### 运行微信小程序

```bash
yarn dev:mp-weixin
# 或者
yarn dev
```

### 打包发行微信小程序

```bash
yarn build:mp-weixin
# 或者
yarn prod  
```

## 文件目录说明

```tree
root
├── /doc/               // 用于存放 cli 工程构建的说明文档 
├── /project-doc/       // 用户存放项目的说明文档目录
├── /public/            // 存放在该文件夹的东西不会被打包影响，而是会原封不动的输出到dist文件夹中
│  └── /index.html      // 入口模板文件
├── /src/               // 存放项目源码及需要引用的资源文件
│  ├── /api/            // 存放项目中接口文件
│  ├── /assets/         // 存放项目中需要用到的资源文件，css、js、images等
│  ├── /components/     // 存放 vue 开发中一些公共组件：header.vue、footer.vue等
│  ├── /directives/     // 全局指令注册
│  ├── /mixins/         // 存放全局混入业务代码
│  ├── /pages/          // 存在vue页面组件的文件夹。
│  ├── /plugins/        // 项目常用的插件集合
│  ├── /service/        // requests 实例配置
│  ├── /router/         // 路由配置
│  ├── /static/         // 存放应用引用静态资源（如图片、视频等）的目录，注意：静态资源只能存放于此
│  ├── /store/          // 存放 vuex 为 vue 专门开发的状态管理器
│  ├── /uni_modules/    // 存放符合 uni-app 的插件模块化规范开发的插件。比如 uView2.0
│  ├── /util/           // 存放 vue 开发过程中一些公共的 js 方法
│  ├── App.vue          // 项目入口文件
│  ├── main.js          // 入口文件
│  ├── manifest.json    // uniapp 管理各个环境下的的配置
│  ├── pages.json       // 页面和路由配置文件
│  └── uni.scss         // uni.scss 是一个特殊文件，在代码中无需 import 这个文件即可在 scss 代码中使用这里的样式变量
├── .editorconfig       // 统一团队代码风格的配置文件
├── .env[.development|production][.local]      // vue-cli 环境变量设置
├── .eslintignore       // eslint 忽略文件配置
├── .eslint             // eslint 文件配置
├── .gitignore          // git 忽略文件配置
├── .npmrc              // npm 镜像配置文件
├── .prettierrc.js      // prettier 统一代码格式化风格配置文件
├── .yarnrc             // yarn 镜像配置文件
├── babel.config.js     // 对 js 文件进行编译转换增强的配置文件
├── jsconfig.json       // JavaScript 语言服务的配置文件。比如：代码提示、文件索引等问题
├── LICENSE             // MIT 许可证
├── package.json        // 包管理配置文件
├── postcss.config.js   // 对 css 文件进行编译转换增强的配置文件
├── README.md           // uni-app-cli-preset 说明文档 
├── vue.config.js       // vue-cli 配置文件
└── yarn.lock           // yarn 锁定依赖版本
```

## 使用到的框架和库

- [vue](https://cn.vuejs.org/)
- [vuex](https://vuex.vuejs.org/zh/)
- [uview](https://www.uviewui.com/)
- [ColorUI](https://github.com/weilanwl/ColorUI)
- [luch-request](https://github.com/lei-mu/luch-request)

## 关键字

- vue （vue 框架）
- uview （ui 组件库）
- vuex （管理器）

## 版本控制

  该项目使用 Git 进行版本管理。

## 作者

- _Cest_
- _wangProgrammer_ itwangyang@foxmail.com

## 鸣谢

- 感谢 [Vue-CLI](https://cli.vuejs.org)
- 感谢 [vue](https://cn.vuejs.org/)
- 感谢 [vuex](https://vuex.vuejs.org/zh/)
- 感谢 [uniapp](https://uniapp.dcloud.io/)
- 感谢 [uview](https://www.uviewui.com/)
- 感谢 [ColorUI](https://github.com/weilanwl/ColorUI)
- 感谢 [luch-request](https://github.com/lei-mu/luch-request)


CHANGELOG.md
<!--
 * @Author: cest
 * @Date: 2022-07-02 12:33:28
 * @LastEditTime: 2022-07-25 07:52:42
 * @LastEditors: cest
 * @FilePath: /uni-app-cli-preset/CHANGELOG.md
 * @Description: 更新日志
-->
# Change Log

## 2022-7-24

更新：

- 【color-ui】
  - 更新按钮说明文档
  - 样式模块更名 line.scss -> cu-box-line.scss
  - 样式模块更名 lines.scss -> cu-box-lines.scss
  - 样式模块更名 box-shadow.scss -> cu-box-shadow.scss
  - 更新标签说明文档

## 2022-7-23 v1.2.0.20220723

累计更新（2022-7-16 ~ 2022-7-23）

主要更新：

- 新增图标字体库 iconfont-24gl-1.0-no-circle-no-squarel、iconfont-uview2
- 修改vuex modules 动态导入的实现方式，解决 modules 不同命名空间下不能重名的问题
- 修改根字体大小 12px
- 更名 colorui -> color-ui

- 更新【color-ui】说明文档与demo
  - 更新辅助布局说明文档
  - 更新背景说明文档
  - 更新文本说明文档
  - 更新图标说明文档

## 2022-7-23

- 删除样式 .test
- 新增样式模块 overflow
- 删除样式模块 cedo-item
- 重构 store string_map 与 types
- 新增图标库： iconfont-24gl-1.0-no-circle-no-squarel
- 新增图标字体库 iconfont-uview2
- 新增主题 变量
- 修改根字体大小 12px
- 删除不必要的文件

## 2022-7-17

- 更名 colorui -> color-ui
- 【color-ui】
  - 更新图标说明文档
  - 更新图标说明文档 优化排版

## 2022-7-16

- 【colorui】
  - 更新辅助布局说明文档
  - 更新背景说明文档
  - 更新文本说明文档

## 2022-7-15

- 【uni-app-cli-preset】 工程
  - 修改vuex modules 动态导入的实现方式，解决 modules 不同命名空间下不能重名的问题
  - scss 混入测试

## 2022-7-14

- 【colorui】
  - 更新 grid 布局 demo和 grid.scss 说明文档

## 2022-7-12 v1.1.0.20220712

更新：

- 累计更新（2022-7-3 v1.0.1-alpha.20220703 ~ 2022-7-10）
- 集成 z-paging

## 2022-7-10

更新：

- 【uni-app-cli-preset】 工程
  - 首页

    ![首页](./docs/screenshots/index_20220710210717.png '首页')
  - 新增 新闻/资讯类 App 模板

    - 新闻-首页

      ![新闻-首页](./docs/screenshots/news_index.png '新闻-首页')

    - 新闻-详情页

      ![新闻-详情页](./docs/screenshots/news_detail.png '新闻-详情页')
  - 修复 vscode 使用了 `@/*` 别名后，路径不提示的问题
  - 取消宽屏适配

## 2022-7-9

更新：

- 【uni-app-cli-preset】 工程
  - 修改 ESlint 规范为 standard
  - 更新说明文档，只保留使用 yarn 的说明
  - 升级脚手架 @dcloudio/uni-* 版本 2.0.0-32920211029004 -> 2.0.1-34920220630001

## 2022-7-8

更新：

- 【uni-app-cli-preset】 工程
  - 动画性能优化
  - ESlint 校验规则 全力全开

## 2022-7-5 v1.0.1-alpha.20220705

更新：

- 【uni-app-cli-preset】 工程
  - 修改 `Vue` 扩展属性的扩展规则
    - `Vue.prototype.StatusBar` -> `Vue.prototype.$StatusBar`
    - `Vue.prototype.Custom` -> `Vue.prototype.$Custom`
    - `Vue.prototype.CustomBar` -> `Vue.prototype.$CustomBar`
- 【colorui】库
  - 新增【Layout】页面
  - 编写 Flex 布局说明文档

## 2022-7-4 v1.0.1-alpha.20220704

更新：

- 【colorui】库
  - 新增【插件扩展】导航页面（组件）

## 2022-7-3 v1.0.1-alpha.20220703

更新：

- 【uni-app-cli-preset】 工程
  - 修复：README.md `更新记录` 链接不正确的问题
  - package.json
    - 新增对 stylus 语法支持
  - 新增配置文件.stylintrc - stylus 语法校验与代码格式风格设置  
- 【colorui】库
  - 新增 CuTabbar 组件
  - 新增【基础元素】导航页面（组件）
  - 新增【交互组件】导航页面（组件）

## 2022-7-3 v1.0.0.20220703

 更新：

- 【uni-app-cli-preset】 工程
  - colorui 移动到 util 目录
  - 修改 .gitignore文件
  - 修改 babel.config.js 文件
  - 修改 README.md 说明文档
  - main.js 删除冗余功能
  - pages.json 精简注释
  - manifest.json 修改微信小程序项目设置
  - package.json 删除不需要的依赖
    - 新增 postcss-comment 依赖项
  的 resolutions 字段用于解析选择性版本，可以 辑 yarn.lock 文件。 详细 <https://  cs/selective-version-resolutions>
    - 升级 vue-cli 版本： 4.5.0 -> 4.5.15
  件的业务处理
  - 目录更名
    - doc -> docs
    - project-doc ->project-docs
  - CHANGELOG.md 从 docs 目录转移到根目录
  - docs 目录：文件名变更 README.md -> index.md

## 2022-7-2

- 【uni-app-cli-preset】 工程
  - 删除无关组件与文件
  - 新增配置文件 `.editorconfig` - 编辑器代码风格配置
  - 新增配置文件 `.env[.*][.local]` - 编译环节配置
  - 新增配置文件 `.npmrc` - npm 镜像配置
  - 新增配置文件 `.prettierrc` - 代码格式化风格配置
  - 新增配置文件 `.yarnrc` - yarn 镜像配置
  - 新增目录 `doc` - 工程说明文档目录
  - 新增目录 `project-doc` - 创建项目说明文档目录
  - 新增目录 `assets` 资源文件目录
  - 新增目录 `assets/theme` 主题样式目录
  - 集成 `colorui` 库
  - 集成 `uview-ui` 框架
  - 集成 `luch-request` 库
  - 集成 `vuex`
  - 更新：工具包 `util`
    - 新增 Array 类的扩展方法
    - 新增常量目录 CONSTANT/
    - 新增 CONSTANT_MESSAGE 消息类
    - 新增 HTTP_RESPONSE_STATUS_CODES 响应状态码
    - 新增 DateUnit 类 - 把时间转换成有单位的字符串
    - 新增函数 bytesToSize - 数据容量单位转换(kb,mb,gb,tb)
    - 新增函数 formatDate - 日期格式化
    - 新增函数 formatLocation - 经纬度格式化
    - 新增函数 formatTime 时间格式化
    - 新增函数 replaceImg 替换图片
    - 新增插件 Handwriting 手写插件
    - 新增目录 uni 常用工具包
    - 新增目录 `regexp` - 正则表达式目录，用于存放常用的正则表达式
    - 新增文件 `index.js` - 用于导出常用函数
- 更新：uni.scss - 新增对 uview2.0 主题变量的支持

## 2022-5-29

- fix: 删除接口的文档

## 2002-5-26

- delete：删除了 unpackage 文件夹
- fix: 修改了测试生产的接口的配置
- feat: 基于 `Vue CLI` 项目脚手架搭建了 uni-app-cli-preset 工程化构建工具。集成 **uni-app + vue-cli + uview2.0 + vue-rouer + vuex + axios + request + tailwindcss + z-paging**


.eslintrc.js
/* eslint-disable no-irregular-whitespace */
/**
 * @Author: cest
 * @Date: 2022-06-17 22:49:05
 * @LastEditTime: 2022-07-09 15:53:12
 * @LastEditors: cest
 * @FilePath: /uni-app-preset/.eslintrc.js
 * @Description: ESLint configuration - 代码质量和纠错配置
 */

// # ESlint使用及配置
module.exports = {
  // ## 配置当前目录为 root
  //
  // ESLint检测配置文件步骤：
  // - 在要检测的文件同一目录里寻找 .eslintrc.* 和 package.json；
  // - 紧接着在父级目录里寻找，一直到文件系统的根目录；
  // - 如果在前两步发现有 root：true 的配置，停止在父级目录中寻找 .eslintrc；
  // - 如果以上步骤都没有找到，则回退到用户主目录 ~/.eslintrc 中自定义的默认配置；

  // > 通常我们都习惯把 ESLint 配置文件放到项目根目录，因此可以为了避免 ESLint 校验的时候往父级目录查找配置文件，所以需要在配置文件中加上 root: true。

  root: true,

  // ## env 环境
  //
  // 一个环境定义了一组预定义的全局变量
  //
  // > 指定不同的环境可以给对应环境下提供预设的全局变量，比如说在 browser 环境下，可以使用 window 全局变量；在 node 环境下，可以使用 process 全局变量等。
  //
  // 可用的环境包括：
  // 可选配置项如下：
  //
  // - browser-浏览器环境中的全局变量。
  // - node-Node.js 全局变量和 Node.js 作用域。
  // - commonjs CommonJS 全局变量和 CommonJS 作用域 (用于 Browserify/WebPack 打包的只在浏览器中运行的代码)。
  // - shared-node-browser Node.js 和 Browser 通用全局变量。
  // - es6 启用除了 modules 以外的所有 ECMAScript 6 特性（该选项会自动设置 ecmaVersion 解析器选项为 6）。
  // - 更多可见[官网](https://eslint.bootcss.com/docs/user-guide/configuring#specifying-environments)
  //
  // 这些环境并不是互斥的，所以可以同时定义多个。可以在源文件里、在配置文件中或使用 命令行 的 --env 选项来指定环境。

  env: {
    // 配置 env 属性，让 eslint 支持到 es6 语法，并且我们设置环境为 browser（浏览器） 或 node
    browser: true,
    node: true,
    es6: true
  },

  // ## parser 解析器及解析参数
  //
  // 解析器将源代码转换为称为抽象语法树（AST）的数据格式，然后，插件使用这种数据格式围绕代码的外观或行为创建称为 lint 规则的断言。
  //
  // ESLint 附带了一个内置的解析器（称为 espree ），因此如果你只编写标准 JavaScript，就不需要自定义解析器了
  //
  // 如果我们想支持非标准 JavaScript 语法，我们需要做的就是为 ESLint 提供一个可供使用的替代解析器。
  // 根据工程的实际情况，对解析器进行设置，比如基于vue框架的工程需要用到能解析vue语法解析器；再比如使用到 TypeScript 的工程，需要配置能解析 TypeScript 的解析器
  //
  // > 这里可以类比 webpack，自身只能处理 js、及 json 文件，css、html 等文件需要通过自定义 loader 处理来理解
  //
  // 常用解析器：
  // - @babel/eslint-parser：一个对 Babel 解析器的包装，使其能够与 ESLint 兼容
  // - @typescript-eslint/parser：TSLint 被弃用后，将 TypeScript 转换成与 estree 兼容的形式，以便在 ESLint 中使用

  parserOptions: {
    parser: 'babel-eslint', // 解析一些最新的 es 语法
    sourceType: 'module' // 模块为 ECMAScript 模块
  },

  // ## extends配置扩展
  //
  // 实际项目中配置 rules 的时候，不可能团队一条一条的去商议配置，太费精力了。通常的做法是使用业内大家普通使用的、遵循的编码规范；然后通过 extends 去引入这些规范。extends 配置的时候接受字符串或者数组：
  // - eslint 开头的：是 ESLint 官方的扩展；常见配置 eslint:recommended、eslint:all
  // - plugin 开头的：是插件类型扩展，可以省略包名的前缀eslint-plugin-，
  //   比如 plugin:vue/essential，实际上使用的是 eslint-config-vue 插件，而 essential 是配置名，可见插件配置说明，如下图，包 eslint-config-vue 中定义了配置参数 essential
  //
  // > extends 可以理解为 rules 的快捷配置字段，一般先配置 extends 为业内通用规范，且数组是后面的会覆盖前面的，并自定义配置 rules 覆盖 extends 中不满足你想要规范的规则

  extends: [
    // ESLint内置的推荐规则，属性启用一系列核心规则，这些规则报告一些常见问题，在 规则页面 中被标记为 [√]
    // 'eslint:recommended'

    // airbnb 规范是最严格的 ESlint 规范，列出下面几点比较明显的区别：
    //
    // 默认必须要分号，而eslint默认不添加分号
    // 不能使用for循环，推荐使用数组自带的 API 完成遍历工作。
    // 当你必须使用函数表达式（或传递一个匿名函数）时，使用箭头函数符号。
    // 除了这些以外，还有更多严格的规则，可以查看 [Airbnb 规范](https://github.com/yuche/javascript)。
    // airbnb 规范用于 react 项目
    // 'airbnb',

    // standard 是基于 ESlint Recommend 衍生出来的更严格的规范。
    // 这个规范和 recommended 大概有 88 处不同，主要是 recommended 很多都是 off, standard 是 error, 比如 单行代码块两边加空格、禁止使用分号结尾。
    'standard',

    'prettier',

    // 选择预设配置 - 基本
    // "plugin:vue/essential",
    // 选择预设配置 - 推荐
    // "plugin:vue/recommended",
    // 选择预设配置 - 强烈推荐
    'plugin:vue/strongly-recommended'
  ],

  // ## rules 规则
  //
  // ESLint 附带有大量的规则，要改变一个规则设置，你必须将规则 ID 设置为下列值之一：
  // "off" 或 0 - 关闭规则
  // "warn" 或 1 - 开启规则，使用警告级别的错误：warn (不会导致程序退出)
  // "error" 或 2 - 开启规则，使用错误级别的错误：error (当被触发的时候，程序会退出)

  rules: {
    // 'vue/attribute-hyphenation': 'off',
    // 'vue/singleline-html-element-content-newline': 'off',
    // 'vue/html-indent': 'off',
    // 'vue/max-attributes-per-line': 'off',
    // 'vue/html-closing-bracket-newline': 'off',
    // 'vue/html-self-closing': 'off',
    // ESLint rules
    // 'no-empty': 'off',
    // 'import/no-extraneous-dependencies': [
    //   'error',
    //   {
    //     devDependencies: true
    //   }
    // ],
    // 'import/no-extraneous-dependencies': 'off',
    // 'global-require': 'off',
    // 'no-underscore-dangle': 'off',
    // 'import/prefer-default-export': 'off',
    // 'no-useless-escape': 'off',
    // 'import/no-unresolved': 'off',
    // 'no-restricted-syntax': 'off',
    // 'consistent-return': 'off',
    'no-param-reassign': [
      'error',
      {
        props: true,
        ignorePropertyModificationsFor: [
          'e', // for e.returnvalue
          'ctx', // for Koa routing
          'req', // for Express requests
          'request', // for Express requests
          'res', // for Express responses
          'response', // for Express responses
          'state' // for vuex state
        ]
      }
    ]
    // 'no-console': 'off',
    // 'no-shadow': 'off',
    // 'prefer-destructuring': 'off',
    // 'max-len': 'off',
    // 'import/no-absolute-path': 'off',
    // 'default-case': 'off',
    // 'no-plusplus': 'off',
    // 'no-use-before-define': 'off',
    // 'array-callback-return': 'off',
    // 'brace-style': 'off',
    // 'no-tabs': 'off',
    // 'guard-for-in': 'off',
    // eqeqeq: 'off',
    // 'import/extensions': 'off',
    // camelcase: 'off',
    // 'import/no-dynamic-require': 'off',
    // 'no-undef': 'off',
    // 'no-unused-vars': 'off',
    // 'no-unused-expressions': 'off',
    // 'func-names': 'off',
    // 'import/no-named-as-default': 'off',
    // 'no-empty-function': 'off'
  },

  globals: {
    uni: true,
    wx: true,
    getCurrentPages: true,
    plus: true,
    weex: true
  }
}

// # 总结： 为什么要用 ESLint？
//
// 1、背景：
//
// 总结为两点：
// - 保持代码风格一致：
//   一千个程序员，就有一千种代码风格。在前端开发中，有几个至今还在争论的代码风格差异。如果一个工程代码风格不一样，可以想象随着项目迭代扩大，简直不忍直视
//   - 单引号还是双引号？
//   - 代码行结束是否需要分号？
//   - 两个空格还是四个空格？
// - 减少代码出错几率：
//   由于 JavaScript 的灵活性，往往一段代码能有多种写法，这时候也会导致协同时差异。并且，有一些写法可能会导致不易发现的 bug，或者这些写法的性能不好，开发时也应该避免
//
// 为了解决这类静态代码问题，每个团队都需要一个统一的 JavaScript 代码规范，团队成员都遵守这份代码规范来编写代码。当然，靠人来保障代码规范是不可靠的，需要有对应的工具来保障，ESLint 就是这个工具
//
// 2、为什么不是 Prettier
//
// Prettier 也可以保证代码风格一致。是的 Prettier 确实可以按照设置的规则对代码进行统一格式化，但是需要明确的一点是，Prettier 只会在格式上对代码进行格式化，一些隐藏的代码质量问题 Prettier 是无法发现的，而 ESLint 可以。
//
// 3、目标
//
// 开发时提示，保存时自动修复，提交时检测
